File: backend/app/schemas/channel.py
"""Pydantic schemas for Channel API."""
from pydantic import BaseModel, Field, ConfigDict
from uuid import UUID
from datetime import datetime


class ChannelBase(BaseModel):
    """Base schema for Channel."""
    name: str = Field(..., min_length=1, max_length=255, description="Channel display name")
    youtube_channel_id: str = Field(..., min_length=1, max_length=50, description="YouTube channel ID")
    thumbnail_url: str | None = Field(None, max_length=500, description="Channel avatar URL")
    description: str | None = Field(None, description="Channel description from YouTube")
    is_hidden: bool = Field(False, description="Hidden channels don't appear in sidebar")


class ChannelUpdate(BaseModel):
    """Schema for updating a channel (only is_hidden can be changed by user)."""
    is_hidden: bool | None = None


class ChannelResponse(BaseModel):
    """Schema for channel response with video count."""
    id: UUID
    user_id: UUID
    youtube_channel_id: str
    name: str
    thumbnail_url: str | None = None
    description: str | None = None
    is_hidden: bool = False
    video_count: int = Field(0, description="Number of videos in this channel")
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/channel.ts
import { z } from 'zod'

/**
 * Zod schema for Channel validation
 * Matches backend API response (UUID and datetime as strings after JSON serialization)
 */
export const ChannelSchema = z.object({
  id: z.string().uuid(),
  user_id: z.string().uuid(),
  youtube_channel_id: z.string(),
  name: z.string(),
  thumbnail_url: z.string().nullable(),
  description: z.string().nullable(),
  is_hidden: z.boolean(),
  video_count: z.number().int().nonnegative(),
  created_at: z.string(),
  updated_at: z.string(),
})

/**
 * Channel type inferred from Zod schema
 */
export type Channel = z.infer<typeof ChannelSchema>

/**
 * Schema for updating a channel (only is_hidden can be changed)
 */
export const ChannelUpdateSchema = z.object({
  is_hidden: z.boolean().optional(),
})

export type ChannelUpdate = z.infer<typeof ChannelUpdateSchema>

/**
 * Array schema for validating multiple channels
 */
export const ChannelsSchema = z.array(ChannelSchema)
